src/wallet/wallet.cpp

```diff
@@ -2497,178 +2403,201 @@ CAmount CWalletObjBase::GetDebit(const isminefilter& filter) const
+        // With certificate it is up to the transaction to tell whether its credit
+        // won't change anymore
+        if (!fCreditCached)
+            nCreditCached = pwallet->GetCredit(*this, ISMINE_SPENDABLE, fCreditCached, /*keepImmatureVoutsOnly*/false);
+
+        credit += nCreditCached;
     }
-    return 0;
+
+    if (filter & ISMINE_WATCH_ONLY) {
+        // Again here, with certificate it is up to the transaction to tell whether its credit
+        // won't change anymore
+        if (!fWatchCreditCached)
+            nWatchCreditCached = pwallet->GetCredit(*this, ISMINE_WATCH_ONLY, fWatchCreditCached, /*keepImmatureVoutsOnly*/false);
+
+        credit += nWatchCreditCached;
+    }
+
+    return credit;
 }
 
 CAmount CWalletObjBase::GetImmatureCredit(bool fUseCache) const
 {
+    if (!IsInMainChain())
+        return CAmount(0);
+
+    if (!IsCoinBase() && !IsCertificate())
+        return CAmount(0);
+
     if (fUseCache && fImmatureCreditCached)
         return nImmatureCreditCached;
-    nImmatureCreditCached = pwallet->GetCredit(*this, ISMINE_SPENDABLE);
-    fImmatureCreditCached = true;
+
+    nImmatureCreditCached = pwallet->GetCredit(*this, ISMINE_SPENDABLE, fImmatureCreditCached, /*keepImmatureVoutsOnly*/true);
     return nImmatureCreditCached;
+
+}
+
+CAmount CWalletObjBase::GetImmatureWatchOnlyCredit(const bool& fUseCache) const
+{
+    if (!IsInMainChain())
+        return CAmount(0);
+
+    if (!IsCoinBase() && !IsCertificate())
+        return CAmount(0);
+
+    if (fUseCache && fImmatureWatchCreditCached)
+        return nImmatureWatchCreditCached;
+
+    nImmatureWatchCreditCached = pwallet->GetCredit(*this, ISMINE_WATCH_ONLY, fImmatureWatchCreditCached, /*keepImmatureVoutsOnly*/true);
+    return nImmatureWatchCreditCached;
 }
-#endif
 
-#if 0
-CAmount CWalletTx::GetAvailableCredit(bool fUseCache) const
-#else
 CAmount CWalletObjBase::GetAvailableCredit(bool fUseCache) const
-#endif
 {
     if (pwallet == 0)
         return 0;
 
-    // Must wait until coinbase is safely deep enough in the chain before valuing it
-    if (IsCoinBase() && !IsMature())
-        return 0;
-
     if (fUseCache && fAvailableCreditCached)
         return nAvailableCreditCached;
 
     CAmount nCredit = 0;
-    uint256 hashTx = GetHash();
-    for (unsigned int i = 0; i < vout.size(); i++)
+    fAvailableCreditCached = true;
+    for (unsigned int pos = 0; pos < vout.size(); pos++)
     {
-        if (!pwallet->IsSpent(hashTx, i))
-        {
-            const CTxOut &txout = vout[i];
-            nCredit += pwallet->GetCredit(txout, ISMINE_SPENDABLE);
+        COutputEntry::maturityState outputMaturity = this->IsOutputMature(pos);
+        if (outputMaturity == COutputEntry::maturityState::NOT_APPLICABLE) {
+            fAvailableCreditCached = false;
+            return CAmount(0);
+        }
+
+        if (outputMaturity == COutputEntry::maturityState::IMMATURE) {
```

Comment:
Would it be better to check against != MATURE? In this case we also would skip the NOT_APPLICBLE. The same for other similar checks

Reply:
`COutputEntry::maturityState::NOT_APPLICABLE` being valid for the tx as a whole, we can have an early return by distinguishing the cases